032-projects/proposal.md

```diff
@@ -0,0 +1,274 @@
+
+Within the `plan`, you may choose to just run a task, using projects as a simple pipeline-less CI flow:
+
+```yaml
+name: ci
+plan:
+- get: booklit
+- task: test
+```
+
+...or you may choose to leverage the `set_pipeline` step to automate your pipeline configuration:
+
+```yaml
+name: ci
+plan:
+- set_pipeline: booklit
+```
+
+New project builds can be triggered by `get` steps with `trigger: true`, however `get` steps cannot specify any `passed` constraints:
+
+```yaml
+name: ci
+plan:
+- get: booklit
+- get: golang
+  trigger: true
+- task: test
+  image: golang
+```
+
+If your pipelines use templating, you can automate that just by using a `task`:
+
+```yaml
+name: ci
+plan:
+- task: generate-pipeline
+- set_pipeline: my-pipeline
+  file: generated-pipeline/pipeline.yml
+```
+
+This versatility should help smoothen Concourse's learning curve - there now exists a simpler CI/CD workflow suitable for side-projects and simple workflows that don't really need pipelines at all, and along the way we've enabled arbitrary automation needs for more complex workflows.
+
+## Project-level resources
+
+Projects define resources under the `resources/` directory. These resources can be used by all pipelines within the project.
+
+Resources defined at the project level can be propagated between pipelines using `passed` constraints. This is not possible with pipeline-local resource definitions.
+
+## Pipeline-local resources
+
+Pipelines can still define their own resources in addition to the project scope.
+
+There is some risk for confusion as developers will have two places to look for resource definitions, but the benefits seem to outweigh the risk:
+
+* Pipeline-local resources can be used to prevent polluting the project-wide resource scope with resources that are only ever useful for a particularly specialized pipeline.
+* Pipeline-local resources are natural fit for pipeline templates which use `((vars))` to configure their resources. For example, a branch pipeline template would have a pipeline-local resource definition for its branch.
+
+Concourse will prevent 'resource shadowing' - any pipeline which defines a local resource matching the name of a project-wide resource should error upon configuration.
+
+## Project-level tasks
+
+Projects define tasks under the `tasks/` directory.
+
+Tasks will automatically propagate to every build plan executed within the project via the project's config directory artifact.
+
+The `task` step will be modified to load the task config corresponding to the task name from the project artifact by default.
+
+For example, the following build plan...:
+
+```yaml
+plan:
+- task: foo
+```
+
+...will be interpreted as this:
+
+```yaml
+plan:
+- task: foo
+  file: project-name/tasks/foo.yml
```

Comment:
I'd suggest the taskdir style: `task: foo` would be interpreted by default as `project-name/tasks/foo/task.yml`.

The basic reasons are that tasks can grow large and complex in their own right, so establishing subdirectories in advance saves messy top-level task directories. It's also easier to do things in bulk (eg, use `xargs` to update multiple task YAMLs simultaneously).